### DISCUSSION TITLE
Korim Bitmap32 expects color channels order to be BGRA?

### DISCUSSION THREAD

**Discussion by ShakedDovrat** [korlibs/korge#487] [Q&A]
Hi,
I am seeing some odd behavior. I'm reading a byte buffer which contains image data into a Korim's Bitmap32. If I write the buffer in RGBA channel order then when I display the resulting image I get an image with the correct content (spacially) but with wrong colors, i.e. wrong channel order. The only way I was able to get a valid image is to write to the buffer in ARGB order, and then read it in reversed order, by using big endian. This means that Bitmap32 expected/reads pixels in BGRA order.
This is strange because the code for the RGBA class clearly states that the order should be RGBA.
This is the code that results in a valid image:
```
// byteBuffer contains pixels in ARGB order
val byteBufferOrdered: ByteBuffer = byteBuffer.order(ByteOrder.BIG_ENDIAN) // big endian -> reversed order
val intBuffer: IntBuffer = byteBufferOrdered.asIntBuffer()
val intArray = IntArray(intBuffer.limit())
intBuffer.get(intArray)
val bitmapKorim = Bitmap32(Video.WIDTH, Video.HEIGHT, RgbaArray(intArray))
// bitmapKorim is a valid image (but shouldn't be...)
```
Now, even though this is fixed, I get a similar issue when converting a pre-populated Android Bitmap, which I cannot reorder w/o a runtime cost:
```
// bitmap is a valid Android Bitmap, acquired from a phone's camera
val bitmapKorim = AndroidNativeImage(bitmap).toBMP32()
// bitmapKorim has wrong color channels
``` 

EDIT: I've managed to reorder the channels, in a way which implies I'm correct about the BGRA:
```
// bitmap is a valid Android Bitmap, acquired from a phone's camera
val bitmapKorim = AndroidNativeImage(bitmap).toBMP32()
val r = bitmapKorim.extractChannel(BitmapChannel.RED)
val b = bitmapKorim.extractChannel(BitmapChannel.BLUE)
bitmapKorim.writeChannel(BitmapChannel.RED, b) // switching red and blue channels
bitmapKorim.writeChannel(BitmapChannel.BLUE, r)
// bitmapKorim is now a valid image
``` 

Please help, thank you

**Comment by soywiz** [comment]
@ShakedDovrat https://github.com/korlibs/korge/commit/bd848cdf418c8d9fae8e7ddf84166090bf8a6c40 This was fixed a couple of days ago, but the latest version published version is from 4 days ago: https://github.com/korlibs/korge-next/releases/tag/v3.0.0-Beta5
Maybe your problem is related to that.

**Reply by ShakedDovrat** [reply]
Thanks. I don't see anything related in the change log
